braindecode/models/eegminer.py

```diff
@@ -0,0 +1,400 @@
+"""
+* Copyright (C) Cogitat, Ltd.
+* Creative Commons Attribution-NonCommercial 4.0 International (CC BY-NC 4.0)
+* Patent GB2609265 - Learnable filters for eeg classification
+* https://www.ipo.gov.uk/p-ipsum/Case/ApplicationNumber/GB2113420.0
+"""
+
+from functools import partial
+
+import torch
+import torch.nn.functional as F
+from einops.layers.torch import Rearrange
+from torch import nn
+from torch.fft import fftfreq
+
+from braindecode.models.base import EEGModuleMixin
+
+
+def torch_hilbert_freq(x, forward_fourier=True):
+    """Computes the Hilbert transform using PyTorch,
+    with the real and imaginary parts as separate dimensions.
+
+    Input shape (forward_fourier=True): (..., seq_len)
+    Input shape (forward_fourier=False): (..., seq_len / 2 + 1, 2)
+    Output shape: (..., seq_len, 2)
+    """
+    if forward_fourier:
+        x = torch.fft.rfft(x, norm=None, dim=-1)
+        x = torch.view_as_real(x)
+    x = x * 2.0
+    x[..., 0, :] = x[..., 0, :] / 2.0  # Don't multiply the DC-term by 2
+    x = F.pad(
+        x, [0, 0, 0, x.shape[-2] - 2]
+    )  # Fill Fourier coefficients to retain shape
+    x = torch.view_as_complex(x)
+    x = torch.fft.ifft(x, norm=None, dim=-1)  # returns complex signal
+    x = torch.view_as_real(x)
+
+    return x
+
+
+def plv_time(x, forward_fourier=True):
+    """Phase Locking Value metric in time domain.
+    x (..., channels, time/(freqs, 2)) -> (..., channels, channels)"""
+    x_a = torch_hilbert_freq(x, forward_fourier)
+    amp = torch.sqrt(x_a[..., 0] ** 2 + x_a[..., 1] ** 2 + 1e-6)
+    x_u = x_a / amp.unsqueeze(-1)
+    x_u_rr = torch.matmul(x_u[..., 0], x_u[..., 0].transpose(-2, -1))
+    x_u_ii = torch.matmul(x_u[..., 1], x_u[..., 1].transpose(-2, -1))
+    x_u_ri = torch.matmul(x_u[..., 0], x_u[..., 1].transpose(-2, -1))
+    x_u_ir = torch.matmul(x_u[..., 1], x_u[..., 0].transpose(-2, -1))
+    r = x_u_rr + x_u_ii
+    i = x_u_ri - x_u_ir
+    time = amp.shape[-1]
+    plv = 1 / time * torch.sqrt(r**2 + i**2 + 1e-6)
+
+    return plv
+
```

Comment:
This implementation is too verbose e overcomplicated, maybe rewrite with Rearrange layers or something else.